Csv modeling

Asmaa7071 opened the pull request:
Pull Request Description:
Overview: This PR introduces functionality to parse and read a CSV file of meal data, convert it into structured Meal objects, and display it through a simple user interface. The primary components introduced in this PR are:

CsvMealRepository: Handles fetching and processing meal data from the CSV file using CsvFileReader and CsvParser.

CsvFileReader: Reads the CSV file, handling any exceptions during the read process.

CsvParser: Converts each row of CSV data into a Meal object, including parsing tags and nutrition information.

FoodChangeMoodUI: A terminal-based user interface that allows users to interact with the meal data by displaying the first 10 meals.

Key Changes:

Meal Data Handling: A new CsvMealRepository has been added that uses a CSV file reader and parser to transform the raw CSV data into Meal objects.

Error Handling: The CsvFileReader and CsvParser components now handle potential errors gracefully by skipping problematic rows and providing relevant warnings.

Dependency Injection: The project now uses Koin for dependency injection, simplifying the management of components like CsvMealRepository, CsvParser, and the UI.

User Interface: A basic CLI user interface (FoodChangeMoodUI) has been added, which allows the user to view meal data interactively.

Testing & Functionality:

Successfully reads and parses meal data from the CSV file.

Handles CSV file errors by skipping faulty rows and logging warnings.

User interface allows displaying the first 10 meals, showing key meal information (e.g., name, ID, nutrition).

How to Test:

Make sure the food.csv file is located in the assets folder.

Run the application and select the option to display the first 10 meals.

Observe the meal data printed on the console.

Next Steps:

Improve user interaction with additional features, such as filtering meals or displaying meal details.

Commit:
data modeling
- create meal data class
- create nutrition data class

AbdoMagdey24 committed:
prepare column index to read from file
and prepare the FoodRepo interface

Commit:
read food csv file
- use library csv reader
- parsing csv data

AbdoMagdey24 committed:
apply dependency injection using koin

AbdoMagdey24 committed:
just for test
- add use case to get first 10 meals
- apply dependency injection for it
- use this use case in the ui

AmeerAmjed commented on src/main/kotlin/logic/GetFirstTenMealsUseCase.kt:
```diff
@@ -0,0 +1,10 @@
+class GetFirstTenMealsUseCase(
+    val mealRepo: MealRepository,
+) {
+    fun getMeals(): List<Meal> = mealRepo.getAllMeals().take(10)
```
Each use case has one functionality, so we can improve this function using the **invoke operator**.

Asmaa7071 replied:
this is for test only

AmeerAmjed commented on src/main/kotlin/logic/GetFirstTenMealsUseCase.kt:
```diff
@@ -0,0 +1,10 @@
+class GetFirstTenMealsUseCase(
+    val mealRepo: MealRepository,
+) {
+    fun getMeals(): List<Meal> = mealRepo.getAllMeals().take(10)
```
We need the size to be flexible and passed as a parameter.

AmeerAmjed commented on src/main/kotlin/model/Nutrition.kt:
```diff
@@ -0,0 +1,11 @@
+package model
+
+data class Nutrition(
+    val calories: Double?,
```
Avoid setting properties as nullable in the model

AmeerAmjed commented on src/main/kotlin/di/AppModule.kt:
```diff
@@ -0,0 +1,21 @@
+import java.io.File
+
+val appModule = module {
+    single { File("assets/food.csv") }
```
Don't hardcode the path; move it to a constants file instead

AmeerAmjed replied:
Why are you using a hardcoded path in two places? If the path needs to change, you'll have to update it in multiple places. We should define it in one place to make future changes easier

AmeerAmjed commented on src/main/kotlin/data/CsvMealRepository.kt:
```diff
@@ -0,0 +1,17 @@
+class CsvMealRepository(private val parser: CsvParser, private val csvFileReader: CsvFileReader) : MealRepository {
+    private val allMeals: MutableList<Meal> = mutableListOf()
+    override fun getAllMeals(): List<Meal> {
+        csvFileReader.readLinesFromFile().map { row ->
```
We don't need a library to read CSV files; we can follow Bareq's video to do that

AmeerAmjed commented on src/main/kotlin/data/CsvFileReader.kt:
```diff
@@ -0,0 +1,31 @@
+class CsvFileReader {
+    fun readLinesFromFile(): List<Map<String, String>> {
+        val mealMap:MutableList<Map<String,String>> = mutableListOf()
+        csvReader().readAllWithHeader(getCsvFile()).map { row ->
```
We don't need a library to read CSV files; we can follow Bareq's video to do that

Asmaa7071 replied:
done

AmeerAmjed commented on src/main/kotlin/data/CsvMealRepository.kt:
```diff
@@ -0,0 +1,17 @@
+import org.damascus.logic.MealRepository
+import org.damascus.model.Meal
+
+class CsvMealRepository(private val parser: CsvParser, private val csvFileReader: CsvFileReader) : MealRepository {
```
Don't use technology names, file extensions, or library names in naming.

AmeerAmjed replied:
Why are we passing CsvParser/CsvFileReader to the repository? We can improve this by moving it to a service or data source layer to make it easier to change the data source in the future

AmeerAmjed commented on src/main/kotlin/model/Meal.kt:
```diff
@@ -0,0 +1,18 @@
+    val steps: List<String>,
+    val description: String,
+    val ingredients: List<String>,
+    val nIngredients: Int?
```
Avoid naming like nSteps (you just added the letter 'n' to make it different from steps). The name should be readable and clear

AmeerAmjed reviewed (changes requested):
Good job! Just a few notes — we need to improve it.

AbdoMagdey24 committed:
resolve comments
- put the file path in constants file instead of hardcode
- use invoke operator in use case

AbdoMagdey24 committed:
delete the companion object to make the file path in one location

AbdoMagdey24 committed:
separate csv reading/parsing logic from CsvMealRepository into CsvMealDataSource
to improve separation of concerns and make future data source changes easier

Commit:
read csv file directly instead of csv reader library

AbdoMagdey24 committed:
add exit option to cli

AmeerAmjed commented on build.gradle.kts:
```diff
@@ -14,6 +14,8 @@ repositories {
 
 dependencies {
     testImplementation(kotlin("test"))
+    implementation("com.github.doyaaaaaken:kotlin-csv-jvm:1.8.0")
```
If you don't use the library, why don't you remove it from Gradle?

Commit:
add utils package

Commit:
Merge remote-tracking branch 'origin/csv-modeling' into csv-modeling
